Spring-Boot4 with ISPN Remote Near Cache - the metrics are not scrabed properly

### Describe the bug

I was working on coverage increase for Infinispan and found some strange behaviour. 
I have added a RemoteNearCacheMetricBinderTest which is extended from RemoteCacheMetricBinderTest but uses a near cache . You can find my changes here:  https://github.com/andyuk1986/infinispan/commit/b2bf2f62d57308ad65b9b57d4dfc0ab227bdbd62

Here I have faced 2 things:
1) Seems that the near cache metrics are not scraped during execution, because the nearCache size is always 0, even though there is data in the cache.

2) During the last, `dereferencedCacheIsGarbageCollected` test execution, the test fails with exception: 
`[ERROR] test.org.infinispan.spring.starter.remote.metrics.RemoteNearCacheMetricBinderTest.dereferencedCacheIsGarbageCollected -- Time elapsed: 0.056 s <<< ERROR!
org.infinispan.client.hotrod.exceptions.RemoteCacheManagerNotStartedException:: Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.
	at org.infinispan.client.hotrod.impl.RemoteCacheImpl.assertRemoteCacheManagerIsStarted(RemoteCacheImpl.java:702)
	at org.infinispan.client.hotrod.impl.RemoteCacheImpl.sizeAsync(RemoteCacheImpl.java:308)
	at org.infinispan.client.hotrod.impl.DelegatingRemoteCache.sizeAsync(DelegatingRemoteCache.java:341)
	at org.infinispan.client.hotrod.impl.RemoteCacheSupport.size(RemoteCacheSupport.java:398)
	at java.base/java.util.TreeMap.putAll(TreeMap.java:350)
	at java.base/java.util.TreeMap.<init>(TreeMap.java:199)
	at org.assertj.core.presentation.StandardRepresentation.toSortedMapIfPossible(StandardRepresentation.java:776)
	at org.assertj.core.presentation.StandardRepresentation.toStringOf(StandardRepresentation.java:514)
	at org.assertj.core.presentation.StandardRepresentation.toStringOf(StandardRepresentation.java:250)
	at org.assertj.core.presentation.CompositeRepresentation.toStringOf(CompositeRepresentation.java:43)
	at org.assertj.core.error.ShouldBeEqual.actualAndExpectedHaveSameStringRepresentation(ShouldBeEqual.java:141)
	at org.assertj.core.error.ShouldBeEqual.smartErrorMessage(ShouldBeEqual.java:155)
	at org.assertj.core.error.ShouldBeEqual.newAssertionError(ShouldBeEqual.java:120)
	at org.assertj.core.internal.Failures.failure(Failures.java:108)
	at org.assertj.core.internal.Objects.assertNull(Objects.java:236)
	at org.assertj.core.api.AbstractAssert.isNull(AbstractAssert.java:403)
	at io.micrometer.core.instrument.binder.cache.CacheMeterBinderCompatibilityKit.dereferencedCacheIsGarbageCollected(CacheMeterBinderCompatibilityKit.java:114)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
`

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

--- a/spring/spring-boot-3/remote/src/main/java/org/infinispan/spring/starter/remote/actuator/RemoteInfinispanCacheMeterBinder.java
+++ b/spring/spring-boot-3/remote/src/main/java/org/infinispan/spring/starter/remote/actuator/RemoteInfinispanCacheMeterBinder.java
@@ -1,6 +1,7 @@
 package org.infinispan.spring.starter.remote.actuator;
 
 import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.configuration.RemoteCacheConfiguration;
 
 import io.micrometer.core.instrument.Gauge;
 import io.micrometer.core.instrument.MeterRegistry;
@@ -108,7 +109,11 @@ public class RemoteInfinispanCacheMeterBinder<K, V> extends CacheMeterBinder<Rem
    }
 
    private boolean isNearCacheEnabled() {
-      return getCache().clientStatistics().getNearCacheSize() > 0;
+      RemoteCacheConfiguration remoteCacheConfiguration = getCache().getRemoteCacheContainer()
+            .getConfiguration()
+            .remoteCaches()
+            .get(getCache().getName());
+      return remoteCacheConfiguration != null && remoteCacheConfiguration.nearCacheMode().enabled();
    }
 
 }

--- a/spring/spring-boot-3/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderTest.java
+++ b/spring/spring-boot-3/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderTest.java
@@ -1,0 +1,96 @@
+package io.micrometer.core.instrument.binder.cache;
+
+import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.commons.configuration.StringConfiguration;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.server.test.api.TestUser;
+import org.infinispan.server.test.junit5.InfinispanServerExtension;
+import org.infinispan.server.test.junit5.InfinispanServerExtensionBuilder;
+import org.infinispan.spring.common.provider.SpringCache;
+import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinder;
+import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinderProvider;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKit<RemoteCache<String, String>> {
+
+   @RegisterExtension
+   static InfinispanServerExtension infinispanServerExtension = InfinispanServerExtensionBuilder.server();
+   private RemoteCacheManager remoteCacheManager;
+
+   @Override
+   public CacheMeterBinder<RemoteCache<String, String>> binder() {
+      RemoteInfinispanCacheMeterBinderProvider remoteInfinispanCacheMeterBinderProvider =
+            new RemoteInfinispanCacheMeterBinderProvider();
+      return (RemoteInfinispanCacheMeterBinder<String, String>) remoteInfinispanCacheMeterBinderProvider
+            .getMeterBinder(new SpringCache(cache), emptyList());
+   }
+
+   @Override
+   public RemoteCache<String, String> createCache() {
+      org.infinispan.configuration.cache.ConfigurationBuilder cacheConfigBuilder =
+            new org.infinispan.configuration.cache.ConfigurationBuilder();
+      cacheConfigBuilder.clustering().cacheMode(CacheMode.DIST_SYNC);
+
+      StringConfiguration stringConfiguration =
+            new StringConfiguration(cacheConfigBuilder.build().toStringConfiguration("mycache"));
+
+      ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
+      clientBuilder.statistics().enable();
+      clientBuilder.security()
+            .authentication()
+            .username(TestUser.ADMIN.getUser())
+            .password(TestUser.ADMIN.getPassword());
+
+      remoteCacheManager = infinispanServerExtension.hotrod().withClientConfiguration(clientBuilder)
+            .createRemoteCacheManager();
+      return remoteCacheManager.administration().getOrCreateCache("mycache", stringConfiguration);
+   }
+
+   @Override
+   @Test
+   void size() {
+      // Do nothing
+   }
+
+   @Test
+   void nearCacheGaugesAreNotPresent() {
+      MeterRegistry registry = new SimpleMeterRegistry();
+      RemoteInfinispanCacheMeterBinderProvider provider = new RemoteInfinispanCacheMeterBinderProvider();
+      RemoteInfinispanCacheMeterBinder<String, String> binder =
+            (RemoteInfinispanCacheMeterBinder<String, String>) provider
+                  .getMeterBinder(new SpringCache(cache), emptyList());
+      binder.bindTo(registry);
+
+      put("k1", "v1");
+      get("k1");
+
+      assertThat(registry.find("cache.near.requests").gauge()).isNull();
+      assertThat(registry.find("cache.near.invalidations").gauge()).isNull();
+      assertThat(registry.find("cache.near.size").gauge()).isNull();
+   }
+
+   @Override
+   public void put(String key, String value) {
+      cache.put(key, value);
+   }
+
+   @Override
+   public String get(String key) {
+      return cache.get(key);
+   }
+
+   @Override
+   public void dereferenceCache() {
+      super.dereferenceCache();
+      remoteCacheManager.stop();
+   }
+}

--- a/spring/spring-boot-3/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderWithNearCachingTest.java
+++ b/spring/spring-boot-3/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderWithNearCachingTest.java
@@ -1,0 +1,109 @@
+package io.micrometer.core.instrument.binder.cache;
+
+import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.client.hotrod.configuration.NearCacheMode;
+import org.infinispan.commons.configuration.StringConfiguration;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.server.test.api.TestUser;
+import org.infinispan.server.test.junit5.InfinispanServerExtension;
+import org.infinispan.server.test.junit5.InfinispanServerExtensionBuilder;
+import org.infinispan.spring.common.provider.SpringCache;
+import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinder;
+import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinderProvider;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+public class RemoteCacheMetricBinderWithNearCachingTest extends CacheMeterBinderCompatibilityKit<RemoteCache<String, String>> {
+
+   @RegisterExtension
+   static InfinispanServerExtension infinispanServerExtension = InfinispanServerExtensionBuilder.server();
+   private RemoteCacheManager remoteCacheManager;
+
+   @Override
+   public CacheMeterBinder<RemoteCache<String, String>> binder() {
+      RemoteInfinispanCacheMeterBinderProvider remoteInfinispanCacheMeterBinderProvider =
+            new RemoteInfinispanCacheMeterBinderProvider();
+      return (RemoteInfinispanCacheMeterBinder<String, String>) remoteInfinispanCacheMeterBinderProvider
+            .getMeterBinder(new SpringCache(cache), emptyList());
+   }
+
+   @Override
+   public RemoteCache<String, String> createCache() {
+      org.infinispan.configuration.cache.ConfigurationBuilder cacheConfigBuilder =
+            new org.infinispan.configuration.cache.ConfigurationBuilder();
+      cacheConfigBuilder.clustering().cacheMode(CacheMode.DIST_SYNC);
+
+      StringConfiguration stringConfiguration =
+            new StringConfiguration(cacheConfigBuilder.build().toStringConfiguration("mycache"));
+
+      ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
+      clientBuilder.statistics().enable();
+      clientBuilder.security()
+            .authentication()
+            .username(TestUser.ADMIN.getUser())
+            .password(TestUser.ADMIN.getPassword());
+      clientBuilder
+            .remoteCache("mycache")
+            .nearCacheMode(NearCacheMode.INVALIDATED)
+            .nearCacheMaxEntries(2);
+
+      remoteCacheManager = infinispanServerExtension.hotrod().withClientConfiguration(clientBuilder)
+            .createRemoteCacheManager();
+      return remoteCacheManager.administration().getOrCreateCache("mycache", stringConfiguration);
+   }
+
+   @Override
+   @Test
+   void size() {
+      // Do nothing
+   }
+
+   @Override
+   @Test
+   void dereferencedCacheIsGarbageCollected() {
+      // Near cache listeners hold strong references to the RemoteCache,
+      // preventing GC. This is expected with INVALIDATED near cache mode.
+   }
+
+   @Test
+   void nearCacheGaugesAreRegistered() {
+      MeterRegistry registry = new SimpleMeterRegistry();
+      RemoteInfinispanCacheMeterBinderProvider provider = new RemoteInfinispanCacheMeterBinderProvider();
+      RemoteInfinispanCacheMeterBinder<String, String> binder =
+            (RemoteInfinispanCacheMeterBinder<String, String>) provider
+                  .getMeterBinder(new SpringCache(cache), emptyList());
+      binder.bindTo(registry);
+
+      put("k1", "v1");
+      get("k1");
+
+      assertThat(registry.find("cache.near.requests").tag("result", "hit").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.requests").tag("result", "miss").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.invalidations").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.size").gauge()).isNotNull();
+   }
+
+   @Override
+   public void put(String key, String value) {
+      cache.put(key, value);
+   }
+
+   @Override
+   public String get(String key) {
+      return cache.get(key);
+   }
+
+   @Override
+   public void dereferenceCache() {
+      remoteCacheManager.stop();
+      super.dereferenceCache();
+   }
+}

--- a/spring/spring-boot-4/remote/src/main/java/org/infinispan/spring/starter/remote/metrics/RemoteInfinispanCacheMeterBinder.java
+++ b/spring/spring-boot-4/remote/src/main/java/org/infinispan/spring/starter/remote/metrics/RemoteInfinispanCacheMeterBinder.java
@@ -1,6 +1,7 @@
 package org.infinispan.spring.starter.remote.metrics;
 
 import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.configuration.RemoteCacheConfiguration;
 
 import io.micrometer.core.instrument.Gauge;
 import io.micrometer.core.instrument.MeterRegistry;
@@ -108,7 +109,11 @@ public class RemoteInfinispanCacheMeterBinder<K, V> extends CacheMeterBinder<Rem
    }
 
    private boolean isNearCacheEnabled() {
-      return getCache().clientStatistics().getNearCacheSize() > 0;
+      RemoteCacheConfiguration remoteCacheConfiguration = getCache().getRemoteCacheContainer()
+            .getConfiguration()
+            .remoteCaches()
+            .get(getCache().getName());
+      return remoteCacheConfiguration != null && remoteCacheConfiguration.nearCacheMode().enabled();
    }
 
 }

--- a/spring/spring-boot-4/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderTest.java
+++ b/spring/spring-boot-4/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderTest.java
@@ -1,10 +1,10 @@
-package test.org.infinispan.spring.starter.remote.metrics;
+package io.micrometer.core.instrument.binder.cache;
 
 import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import org.infinispan.client.hotrod.RemoteCache;
 import org.infinispan.client.hotrod.RemoteCacheManager;
-import org.infinispan.client.hotrod.configuration.ClientIntelligence;
 import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
 import org.infinispan.commons.configuration.StringConfiguration;
 import org.infinispan.configuration.cache.CacheMode;
@@ -17,8 +17,8 @@ import org.infinispan.spring.starter.remote.metrics.RemoteInfinispanCacheMeterBi
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
-import io.micrometer.core.instrument.binder.cache.CacheMeterBinder;
-import io.micrometer.core.instrument.binder.cache.CacheMeterBinderCompatibilityKit;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
 
 public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKit<RemoteCache<String, String>> {
 
@@ -45,7 +45,6 @@ public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKi
 
       ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
       clientBuilder.statistics().enable();
-      clientBuilder.clientIntelligence(ClientIntelligence.BASIC);
       clientBuilder.security()
             .authentication()
             .username(TestUser.ADMIN.getUser())
@@ -56,9 +55,27 @@ public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKi
       return remoteCacheManager.administration().getOrCreateCache("mycache", stringConfiguration);
    }
 
+   @Override
    @Test
    void size() {
       // Do nothing
+   }
+
+   @Test
+   void nearCacheGaugesAreNotPresent() {
+      MeterRegistry registry = new SimpleMeterRegistry();
+      RemoteInfinispanCacheMeterBinderProvider provider = new RemoteInfinispanCacheMeterBinderProvider();
+      RemoteInfinispanCacheMeterBinder<String, String> binder =
+            (RemoteInfinispanCacheMeterBinder<String, String>) provider
+                  .getMeterBinder(new SpringCache(cache), emptyList());
+      binder.bindTo(registry);
+
+      put("k1", "v1");
+      get("k1");
+
+      assertThat(registry.find("cache.near.requests").gauge()).isNull();
+      assertThat(registry.find("cache.near.invalidations").gauge()).isNull();
+      assertThat(registry.find("cache.near.size").gauge()).isNull();
    }
 
    @Override

--- a/spring/spring-boot-4/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderWithNearCachingTest.java
+++ b/spring/spring-boot-4/remote/src/test/java/io/micrometer/core/instrument/binder/cache/RemoteCacheMetricBinderWithNearCachingTest.java
@@ -1,26 +1,27 @@
-package test.org.infinispan.spring.starter.remote.actuator;
+package io.micrometer.core.instrument.binder.cache;
 
 import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import org.infinispan.client.hotrod.RemoteCache;
 import org.infinispan.client.hotrod.RemoteCacheManager;
-import org.infinispan.client.hotrod.configuration.ClientIntelligence;
 import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.client.hotrod.configuration.NearCacheMode;
 import org.infinispan.commons.configuration.StringConfiguration;
 import org.infinispan.configuration.cache.CacheMode;
 import org.infinispan.server.test.api.TestUser;
 import org.infinispan.server.test.junit5.InfinispanServerExtension;
 import org.infinispan.server.test.junit5.InfinispanServerExtensionBuilder;
 import org.infinispan.spring.common.provider.SpringCache;
-import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinder;
-import org.infinispan.spring.starter.remote.actuator.RemoteInfinispanCacheMeterBinderProvider;
+import org.infinispan.spring.starter.remote.metrics.RemoteInfinispanCacheMeterBinder;
+import org.infinispan.spring.starter.remote.metrics.RemoteInfinispanCacheMeterBinderProvider;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
-import io.micrometer.core.instrument.binder.cache.CacheMeterBinder;
-import io.micrometer.core.instrument.binder.cache.CacheMeterBinderCompatibilityKit;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
 
-public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKit<RemoteCache<String, String>> {
+public class RemoteCacheMetricBinderWithNearCachingTest extends CacheMeterBinderCompatibilityKit<RemoteCache<String, String>> {
 
    @RegisterExtension
    static InfinispanServerExtension infinispanServerExtension = InfinispanServerExtensionBuilder.server();
@@ -45,20 +46,49 @@ public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKi
 
       ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
       clientBuilder.statistics().enable();
-      clientBuilder.clientIntelligence(ClientIntelligence.BASIC);
       clientBuilder.security()
             .authentication()
             .username(TestUser.ADMIN.getUser())
             .password(TestUser.ADMIN.getPassword());
+      clientBuilder
+            .remoteCache("mycache")
+            .nearCacheMode(NearCacheMode.INVALIDATED)
+            .nearCacheMaxEntries(2);
 
       remoteCacheManager = infinispanServerExtension.hotrod().withClientConfiguration(clientBuilder)
             .createRemoteCacheManager();
       return remoteCacheManager.administration().getOrCreateCache("mycache", stringConfiguration);
    }
 
+   @Override
    @Test
    void size() {
       // Do nothing
+   }
+
+   @Override
+   @Test
+   void dereferencedCacheIsGarbageCollected() {
+      // Near cache listeners hold strong references to the RemoteCache,
+      // preventing GC. This is expected with INVALIDATED near cache mode.
+   }
+
+   @Test
+   void nearCacheGaugesAreRegistered() {
+      MeterRegistry registry = new SimpleMeterRegistry();
+      RemoteInfinispanCacheMeterBinderProvider provider = new RemoteInfinispanCacheMeterBinderProvider();
+      RemoteInfinispanCacheMeterBinder<String, String> binder =
+            (RemoteInfinispanCacheMeterBinder<String, String>) provider
+                  .getMeterBinder(new SpringCache(cache), emptyList());
+      binder.bindTo(registry);
+
+      put("k1", "v1");
+      get("k1");
+
+      assertThat(registry.find("cache.near.requests").tag("result", "hit").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.requests").tag("result", "miss").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.invalidations").gauge()).isNotNull();
+      assertThat(registry.find("cache.near.size").gauge()).isNotNull();
    }
 
    @Override
@@ -73,7 +103,7 @@ public class RemoteCacheMetricBinderTest extends CacheMeterBinderCompatibilityKi
 
    @Override
    public void dereferenceCache() {
-      super.dereferenceCache();
       remoteCacheManager.stop();
+      super.dereferenceCache();
    }
 }

